src/tests/networking_tests.rs

```diff
@@ -2421,9 +2421,348 @@ pub mod net_tests {
+        //     -(Errno::ENOTCONN as i32)
+        // );
+
+        assert_eq!(cage.bind_syscall(sockfd, &socket), 0);
+        assert_eq!(cage.listen_syscall(sockfd, 10), 0);
+
+        // wrong how argument error
+        assert_eq!(
+            cage.netshutdown_syscall(sockfd, 10),
+            -(Errno::EINVAL as i32)
+        );
+
+        assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        lindrustfinalize();
+    }
+
+    #[test]
+    pub fn ut_lind_net_shutdown_unix() {
+        // this test is used for testing shutdown with UNIX socket
+
+        // acquiring a lock on TESTMUTEX prevents other tests from running concurrently,
+        // and also performs clean env setup
+        let _thelock = setup::lock_and_init();
+
+        let cage = interface::cagetable_getref(1);
+
+        let serverfd = cage.socket_syscall(AF_UNIX, SOCK_STREAM, 0);
+
+        let serveraddr = interface::new_sockaddr_unix(AF_UNIX as u16, "server_shutdown".as_bytes());
+        let serversocket = interface::GenSockaddr::Unix(serveraddr);
+
+        assert_eq!(cage.bind_syscall(serverfd, &serversocket), 0);
+        assert_eq!(cage.listen_syscall(serverfd, 10), 0);
+
+        assert_eq!(cage.fork_syscall(2), 0);
+        assert_eq!(cage.fork_syscall(3), 0);
+
+        let barrier = Arc::new(Barrier::new(2));
+        let barrier_clone = barrier.clone();
+
+        let barrier_shut = Arc::new(Barrier::new(2));
+        let barrier_shut_clone = barrier_shut.clone();
+
+        let thread = interface::helper_thread(move || {
+            let cage2 = interface::cagetable_getref(2);
+
+            let fd = cage2.socket_syscall(AF_UNIX, SOCK_STREAM, 0);
+            assert_eq!(cage2.connect_syscall(fd, &serversocket), 0);
+
+            // first make sure send and recv are working before shutdown
+            assert_eq!(cage2.send_syscall(fd, str2cbuf("client send"), 11, 0), 11);
+            let mut buf = sizecbuf(11);
+            assert_eq!(cage2.read_syscall(fd, buf.as_mut_ptr(), 11), 11);
+            assert_eq!(cbuf2str(&buf), "server send");
+
+            barrier_clone.wait();
+
+            // now let's shutdown RD
+            assert_eq!(cage2.netshutdown_syscall(fd, SHUT_RD), 0);
+
+            // BUG: read after SHUT_RD should not raise ENOTCONN error
+            // the desired behavior is to return end-of-file (0)
+            // assert_eq!(cage2.read_syscall(fd, buf.as_mut_ptr(), 8), 0);
+
+            barrier_shut_clone.wait();
+
+            // write should succeed
+            assert_eq!(
+                cage2.send_syscall(fd, str2cbuf("before SHUT_WR"), 14, 0),
+                14
+            );
+            assert_eq!(cage2.netshutdown_syscall(fd, SHUT_WR), 0);
+
+            assert_eq!(cage2.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        });
+
+        let mut sockgarbage = interface::GenSockaddr::V4(interface::SockaddrV4::default());
+        let fd = cage.accept_syscall(serverfd, &mut sockgarbage);
```

Comment:
how does the following accept work if the client's domain is AF_UNIX?

Reply:
sockgarbage is client's address. Although client's address is unix type, and sockgarbage is V4 type, they are all GenSockaddr, so the client's address could be assigned to sockgarbage. This is how enum works.